python_modules/dagster/dagster/_core/definitions/job_definition.py

```diff
@@ -1012,3 +1013,52 @@ def get_run_config_schema_for_job(
         .get_run_config_schema("default")
         .run_config_schema_type
     )
+
+
+def _infer_asset_layer_from_source_asset_deps(job_graph_def: GraphDefinition) -> AssetLayer:
+    """
+    For non-asset jobs that have some inputs that are fed from SourceAssets, constructs an
+    AssetLayer that includes those SourceAssets.
+    """
+    asset_keys_by_node_input_handle: Dict[NodeInputHandle, AssetKey] = {}
+    source_assets_list = []
+    source_asset_keys_set = set()
+    io_manager_keys_by_asset_key: Mapping[AssetKey, str] = {}
+
+    stack: List[Tuple[GraphDefinition, Optional[NodeHandle]]] = [(job_graph_def, None)]
```

Comment:
a comment here or a named tuple with named properties (just for documentation) saying that each stack entry is a graph_def and its parent node if it not the root. I searched and line 1031 answer the question for me because the variable names of the tuple unpacking. this is ok for now because it is so close, but code gets moved around etc.